tests/services-lorri.nix

```diff
@@ -6,11 +6,19 @@ let
   actual = pkgs.runCommand "convert-plist-to-json" { buildInputs = [ pkgs.xcbuild ]; }
     "plutil -convert json -o $out ${plistPath}";
   actualJson = builtins.fromJSON (builtins.readFile "${actual.out}");
+  expectedPathItems = map (x: "${x}/bin")
```

Comment:
I'm not sure if this is the best way to get this done, I'm still learning nix

Reply:
Yeah, this test is is a bit overly specific.

No need to discard the string context tho since the packages will be referenced by the build anyway. To get rid of the dependencies in the test an overlay is needed in this case, but unlike eg. lorri they're pretty standard in this case so it's not a big deal. I mostly use this to avoid having to download a bunch of unnecessary stuff for the tests or in case the package in nixpkgs breaks I don't really want that to influence these tests.

```nix
{
  nixpkgs.overlays = [
    (self: super:
    {
      git = super.runCommand "git-0.0.0" {} "mkdir $out";
      gnutar = super.runCommand "gnutar-0.0.0" {} "mkdir $out";
      gzip = super.runCommand "gzip-0.0.0" {} "mkdir $out";
    })
  ];
}
```

Reply:
thanks for the feedback! I tried adding the `nixpkgs.overlays`section to the expression at [1] while removing the calls to `builtins.unsafeDiscardStringContext`, but I got an error related to `tar` [2]. I probably need to place the overlays somewhere else, but I'm a bit lost now.

[1] https://github.com/LnL7/nix-darwin/blob/ee9e1205e7892cb012a276ac158600d987cec815/tests/services-lorri.nix#L36-L42

[2]
```
builder for '/nix/store/0rz1d1d2qfy17y37gflkp0pswpp565fm-perl-5.32.0.drv' failed with exit code 1; last 4 log lines:
  unpacking sources
  unpacking source archive /nix/store/bl6b2p0aw8phljjn0s3xzyghbbzawl2r-perl-5.32.0.tar.gz
  /nix/store/9zhvxgmqwph7cd2mc9fwg9y39nziqdaz-stdenv-darwin/setup: line 847: tar: command not found
  do not know how to unpack source archive /nix/store/bl6b2p0aw8phljjn0s3xzyghbbzawl2r-perl-5.32.0.tar.gz
```

Reply:
Yeah this was more an example for reference. These dependencies specifically are very low level and used by many other things so actually replacing them breaks various things like you noticed.

Reply:
gotcha! is it fine then to keep the `builtins.unsafeDiscardStringContext` calls? is there anything else you'd like to see on this pr?

Reply:
No neither the discard context or overlays are needed.

Reply:
hm if I remove the discard context I get this error:
```
nix-repl> :b tests.services-lorri
error: the string '{
  "EnvironmentVariables": {
          "NIX_PATH": "nixpkgs=/nix/store/12pl142h03dcnphqmf1qq1v7pdv12pck-nixpkgs-21.03pre249027.f4254fba4f2/nixpkgs",
          "PATH": "/nix/store/7srl5rd0p8svg4mfwsqc118pxlwx9204-nix-2.3.8/bin:/nix/store/psvzvnzfcsdrlaibzsv7s9wp0ni96978-git-2.28.0/bin:/nix/store/56dw9pwdma8iqxx4a7ibi4l72mcm0px9-gnutar-1.32/bin:/nix/store/y64kgdcxycv5dy62jwffyyh20306xpfr-gzip-1.10/bin"
  },
  "KeepAlive": true,
  "Label": "org.nixos.lorri",
  "ProcessType": "Background",
  "ProgramArguments": [
          "/bin/sh",
          "-c",
          "exec /nix/store/pr02204rvl68n7zggz4jag8ya6jhaaml-lorri-1.2/bin/lorri daemon"
  ],
  "RunAtLoad": true
}
' is not allowed to refer to a store path (such as '!out!/nix/store/49wqxqnnp0aqz7s8ipcnd8xwhbr0bn44-gnutar-1.32.drv'), at /Users/mal/git/nix-darwin/tests/services-lorri.nix:17:18
```

Reply:
Oh I see now, this was there already. Using an attrset instead of going through fromJSON would fix that or just simply checking some specific keys in the build with eg. `jq`. But it's ok to leave that as is if you don't want to spend time on that.

Reply:
> Oh I see now, this was there already
Exactly, it was there before :)

> it's ok to leave that as is if you don't want to spend time on that.
I'm more interested in getting it right :)

> Using an attrset instead of going through fromJSON would fix that

However, I got a bit stuck trying to avoid the `fromJSON` conversion. If I use an attrset and then convert that with `builtins.toJSON` to compare with the actual value, it turns out that the `toJSON` conversion only goes one level deep.

> simply checking some specific keys in the build with eg. jq

This made me think that perhaps the way is avoiding JSON completely and working directly with the plist, but it seems there's very little support for that. The `plutil` command is fairly limited, it seems to lack a simple way to read out a key. I also tried `defaults read`, but it doesn't work with nested dictionaries. My last hope was `PlistBuddy`, which I learned about online. It seems to be part of macOS but it's absent from the PATH; it needs to be run directly as `/usr/libexec/PlistBuddy`. This one is very complete, but I'm not sure if there's a "nix way" to invoke it, and I couldn't find it under `pkgs`

Reply:
I would suggest moving the checks to the build instead like most of the other tests do, but unlike there use the plutil to check specific keys instead just using grep like the other cases do.  Checking the entire content is a bit overkill anyway.

So for example something like this:

```sh
# ...
{ 
   services.lorri.enable = true; 
   test = ''
     export PATH=${lib.makeBinPath [ pkgs.xcbuild pkgs.jq ]}
     plutil -convert json -o service.json ${service}
     [ $(jq -r '.ProgramArguments[-1]' service.json) = "exec ${pkgs.lorri}/bin/lorri daemon" ]
   ''; 
} 
```

Reply:
aha! that makes a lot of sense, thanks! I'll work on it later today

Reply:
done in d037dfa :) please let me know what you think. Thanks for the help!

Reply:
That's great, thanks!